Incorrect behaviour of `prune --production` command

<!--
**Please only file bugs/feature requests for pnpm here.**

If you have a question that is not answered in our FAQ: https://pnpm.js.org/en/faq
then it is best to ask it in our Discord chatroom:
https://discord.gg/mThkzAT

If your issue is a bug, please follow the format below:
-->

### pnpm version: 5.15.1

### Code to reproduce the issue:

- have a package inside `optinalDependencies`
- run `pnpm install --no-optional`
- await installation
- the optional dependencies is correctly NOT installed
- run `pnpm prune --production`

<!--
If there was a fatal error also include a gist of your node_modules/.pnpm-debug.log file.
-->

### Expected behavior: 

`pnpm` should not attempt a make a request to an optional dependency that is not installed.

### Actual behavior: 

`pnpm` tries to make a request to the package inside the `optionalDependencies`.

```
Step 38/40 : RUN pnpm prune --production
 ---> Running in 11d2d5c7c643
Lockfile is up-to-date, resolution step is skipped
Packages: -660
--------------------------------------------------------------------------------
Progress: resolved 1, reused 0, downloaded 0, added 0
 WARN  GET npm.pkg.github.com/download/****/****/1.2.0/*** error (ERR_PNPM_FETCH_401). Will retry in 10 seconds. 2 retries left.
 ERROR  GET npm.pkg.github.com/download/****/****/1.2.0/***: Unauthorized - 401
No authorization header was set for the request.

``` 


The reason why there's 403, is because we have an issue in our environment, that's why we decided to move some dep as optional. This is not happening with `npm`


### Additional information:

 - `node -v` prints: v14.7.0
 - Windows, macOS, or Linux?: Linux

--- a/.changeset/empty-snails-grin.md
+++ b/.changeset/empty-snails-grin.md
@@ -1,0 +1,5 @@
+---
+"@pnpm/headless": patch
+---
+
+A failing optional dependency should not cause a crash of headless installation.

--- a/packages/headless/src/index.ts
+++ b/packages/headless/src/index.ts
@@ -50,6 +50,7 @@ import packageIsInstallable from '@pnpm/package-is-installable'
 import { fromDir as readPackageFromDir } from '@pnpm/read-package-json'
 import { readProjectManifestOnly, safeReadProjectManifestOnly } from '@pnpm/read-project-manifest'
 import {
+  FetchPackageToStoreFunction,
   PackageFilesResponse,
   StoreController,
 } from '@pnpm/store-controller-types'
@@ -592,13 +593,19 @@ async function lockfileToDepGraph (
           requester: opts.lockfileDir,
           status: 'resolved',
         })
-        let fetchResponse = opts.storeController.fetchPackage({
-          force: false,
-          lockfileDir: opts.lockfileDir,
-          pkgId: packageId,
-          resolution,
-        })
-        if (fetchResponse instanceof Promise) fetchResponse = await fetchResponse
+        let fetchResponse!: ReturnType<FetchPackageToStoreFunction>
+        try {
+          fetchResponse = opts.storeController.fetchPackage({
+            force: false,
+            lockfileDir: opts.lockfileDir,
+            pkgId: packageId,
+            resolution,
+          })
+          if (fetchResponse instanceof Promise) fetchResponse = await fetchResponse
+        } catch (err) {
+          if (pkgSnapshot.optional) return
+          throw err
+        }
         fetchResponse.files() // eslint-disable-line
           .then(({ fromStore }) => {
             progressLogger.debug({
@@ -741,7 +748,13 @@ function linkAllPkgs (
 ) {
   return Promise.all(
     depNodes.map(async (depNode) => {
-      const filesResponse = await depNode.fetchingFiles()
+      let filesResponse!: PackageFilesResponse
+      try {
+        filesResponse = await depNode.fetchingFiles()
+      } catch (err) {
+        if (depNode.optional) return
+        throw err
+      }
 
       const { importMethod, isBuilt } = await storeController.importPackage(depNode.dir, {
         filesResponse,

--- a/packages/headless/test/fixtures/has-nonexistent-optional-dep/package.json
+++ b/packages/headless/test/fixtures/has-nonexistent-optional-dep/package.json
@@ -1,0 +1,14 @@
+{
+  "name": "has-nonexistent-optional-dep",
+  "version": "1.0.0",
+  "dependencies": {
+    "is-positive": "^1.0.0",
+    "rimraf": "^2.6.2"
+  },
+  "devDependencies": {
+    "is-negative": "^2.1.0"
+  },
+  "optionalDependencies": {
+    "@zkochan/not-exists": "1.2.0"
+  }
+}

--- a/packages/headless/test/index.ts
+++ b/packages/headless/test/index.ts
@@ -233,6 +233,28 @@ test('not installing optional deps', async () => {
   await project.has('pkg-with-good-optional')
 })
 
+test('skipping optional dependency if it cannot be fetched', async () => {
+  const prefix = path.join(fixtures, 'has-nonexistent-optional-dep')
+  const reporter = sinon.spy()
+
+  await headless(await testDefaults({
+    lockfileDir: prefix,
+    reporter,
+  }, {
+    retry: {
+      retries: 0,
+    },
+  }))
+
+  const project = assertProject(prefix)
+  expect(project.requireModule('is-positive')).toBeTruthy()
+  expect(project.requireModule('rimraf')).toBeTruthy()
+  expect(project.requireModule('is-negative')).toBeTruthy()
+
+  expect(await project.readCurrentLockfile()).toBeTruthy()
+  expect(await project.readModulesManifest()).toBeTruthy()
+})
+
 test('run pre/postinstall scripts', async () => {
   const prefix = path.join(fixtures, 'deps-have-lifecycle-scripts')
   const outputJsonPath = path.join(prefix, 'output.json')

